logstash-core/src/test/java/org/logstash/FileLockFactoryTest.java

```diff
@@ -88,4 +91,50 @@ public void ReleaseUnobtainedLock() throws IOException {
         FileLockFactory.getDefault().releaseLock(lock);
         FileLockFactory.getDefault().releaseLock(lock);
     }
+
+    @Test
+    public void crossJvmObtainLockOnLocked() throws IOException, ClassNotFoundException {
+        // Timeout (in ms).
+        long timeout = 30000;
+        Process p = null;
+        String lockFile = ".testCrossJvm";
+        FileLock lock = null;
+
+        // Build the command to spawn a children JVM.
+        String[] cmd = {
+            Paths.get(System.getProperty("java.home"), "bin", "java").toString(),
+            "-cp", System.getProperty("java.class.path"),
+            Class.forName("org.logstash.FileLockFactoryMain").getName(),
+            lockDir, lockFile
+        };
+
+        try {
+            // Start the children program that will lock the file.
+            p = new ProcessBuilder(cmd).start();
+            InputStream is = p.getInputStream();
+            long endTime = System.currentTimeMillis() + timeout;
+            long now;
+            do {
+                // Wait on children program to write something to stdout.
+                now = System.currentTimeMillis();
+            } while (is.available() == 0 && now < endTime);
+
+            // Check it did not timeout.
+            assertTrue(now < endTime);
+            // Check the children process is still running.
+            assertThat(p.isAlive(), is(equalTo(true)));
+
+            try {
+                // Try to obtain the lock held by the children process.
+                FileLockFactory.getDefault().obtainLock(lockDir, lockFile);
+                fail("Should have threw an exception");
+            } catch (LockException e) {
+                // Expected exception as the file is already locked.
+            }
+        } finally {
+            if (p != null) {
+                p.destroy();
```

Comment:
@wainersm I admit I'm not an expert on the process API but maybe we can make this a little safer using `destroyForcibly` instead of only `destroy`?
In this case the only way we really get a "zombie" would be the initial JVM dying I guess?